pages/api/roles/index.ts

```diff
@@ -1,5 +1,129 @@
+import {
+  createJsonResponse,
+  HTTP_GET_METHOD,
+  HTTP_POST_METHOD,
+  HTTP_STATUS_BAD_REQUEST,
+  HTTP_STATUS_CREATED,
+  HTTP_STATUS_OK,
+  rejectHttpMethod,
+} from '../../../utils/http/httpHelper';
+import { createIfPossible } from '../../../utils/prisma/prismaHelpers';
 
-export default function handler(req: NextApiRequest, res: NextApiResponse) {
-  res.status(200).end(`Request sent to ${req.url}\n`);
+const prisma = new PrismaClient();
+
+enum MessageType {
+  INVALID_COMPANY_ID,
+  COMPANY_DOES_NOT_EXISTS,
+  EMPTY_TITLE,
+  INVALID_TYPE,
+  INVALID_YEAR,
+  ROLE_CREATED_SUCCESSFULLY,
+}
+
+const messages = new Map<MessageType, StatusMessage[]>([
+  [
+    MessageType.INVALID_COMPANY_ID,
+    [{ type: StatusMessageType.Error, message: 'The company for this role is invalid.' }],
+  ],
+  [
+    MessageType.COMPANY_DOES_NOT_EXISTS,
+    [{ type: StatusMessageType.Error, message: 'The company for this role does not exists.' }],
+  ],
+  [MessageType.EMPTY_TITLE, [{ type: StatusMessageType.Error, message: 'Role title is empty.' }]],
+  [MessageType.INVALID_TYPE, [{ type: StatusMessageType.Error, message: 'Role type is invalid.' }]],
+  [MessageType.INVALID_YEAR, [{ type: StatusMessageType.Error, message: 'Role year is invalid.' }]],
+  [
+    MessageType.ROLE_CREATED_SUCCESSFULLY,
+    [{ type: StatusMessageType.Success, message: 'Role was created succesfully.' }],
+  ],
+]);
+
+function handler(req: NextApiRequest, res: NextApiResponse) {
+  const method = req.method;
+
+  switch (method) {
+    case HTTP_GET_METHOD:
+      handleGet(req, res);
+      break;
+    case HTTP_POST_METHOD:
+      handlePost(req, res);
+      break;
+    default:
+      rejectHttpMethod(res, method);
+  }
+}
+
+// TODO: Optional Search query. Will be done in another PR to unblock others.
+
+async function handleGet(_: NextApiRequest, res: NextApiResponse<ApiResponse<RoleListData[]>>) {
+  const roles: RoleListData[] = await prisma.role.findMany({
+    select: {
+      id: true,
+      title: true,
+      type: true,
+      year: true,
+      isVerified: true,
+      company: { select: { id: true, name: true, companyUrl: true } },
+    },
+  });
+
+  res.status(HTTP_STATUS_OK).json(createJsonResponse(roles));
 }
+
+async function handlePost(req: NextApiRequest, res: NextApiResponse<ApiResponse<RoleData | EmptyPayload>>) {
+  if (!(await isValidRequest(req, res))) {
+    return;
+  }
+
+  const rolePostData: RolePostData = req.body;
+
+  createIfPossible(res, async () => {
+    const newRole: RoleData = await prisma.role.create({
+      data: rolePostData,
+      select: { id: true, title: true, type: true, year: true },
+    });
+
+    res
+      .status(HTTP_STATUS_CREATED)
+      .json(createJsonResponse(newRole, messages.get(MessageType.ROLE_CREATED_SUCCESSFULLY)));
+  });
+}
+
+async function isValidRequest(req: NextApiRequest, res: NextApiResponse<ApiResponse<EmptyPayload>>): Promise<boolean> {
```

Comment:
Mentioned elsewhere, prefer handling res inside original + isValidRequest sounds like it only checks and returns boolean, not make destructive actions.